app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectManager.java

```diff
@@ -0,0 +1,489 @@
+      @Override
+      public void onSuccess() {
+        // UI updated from broadcast receiver
+      }
+
+      @Override
+      public void onFailure(int reason) {
+        String errorMessage = getErrorMessage(reason);
+        Log.d(TAG, activity.getString(R.string.connection_failed) + ": " + errorMessage);
+        displayToast(R.string.connection_failed, Toast.LENGTH_LONG);
+      }
+    });
+  }
+
+  public void performHandshakeWithSelectedPeerDevice() {
+    if (BuildConfig.DEBUG) {
+      Log.d(TAG, "Starting handshake");
+    }
+    peerGroupHandshakeAsyncTask = new PeerGroupHandshakeAsyncTask(this);
+    peerGroupHandshakeAsyncTask.execute();
+  }
+
+  public boolean isFileSender() {
+    return isFileSender;
+  }
+
+  public @NonNull ArrayList<Uri> getFileUriArrayList() {
+    return fileUriArrayList;
+  }
+
+  public int getTotalFilesForTransfer() {
+    return totalFilesForTransfer;
+  }
+
+  public void setTotalFilesForTransfer(int totalFilesForTransfer) {
+    this.totalFilesForTransfer = totalFilesForTransfer;
+  }
+
+  public @NonNull ArrayList<FileItem> getFilesForTransfer() {
+    return filesForTransfer;
+  }
+
+  public void setFilesForTransfer(@NonNull ArrayList<FileItem> fileItems) {
+    this.filesForTransfer = fileItems;
+  }
+
+  public void incrementTotalFilesSent() {
+    this.filesSent++;
+  }
+
+  public boolean allFilesSent() {
+    return (filesSent == totalFilesForTransfer);
+  }
+
+  public @NonNull String getZimStorageRootPath() {
+    return (sharedPreferenceUtil.getPrefStorage() + "/Kiwix/");
+  }
+
+  public @NonNull InetAddress getFileReceiverDeviceAddress() {
+    return fileReceiverDeviceAddress;
+  }
+
+  public static void copyToOutputStream(@NonNull InputStream inputStream, @NonNull
+      OutputStream outputStream)
+      throws IOException {
+    byte[] bufferForBytes = new byte[1024];
+    int bytesRead;
+
+    Log.d(TAG, "Copying to OutputStream...");
+    while ((bytesRead = inputStream.read(bufferForBytes)) != -1) {
+      outputStream.write(bufferForBytes, 0, bytesRead);
+    }
+
+    outputStream.close();
+    inputStream.close();
+    Log.d(LocalFileTransferActivity.TAG, "Both streams closed");
+  }
+
+  public void setClientAddress(@Nullable InetAddress clientAddress) {
+    if (clientAddress == null) {
+      // null is returned only in case of a failed handshake
+      displayToast(R.string.device_not_cooperating, Toast.LENGTH_LONG);
+      ((Callbacks) activity).onFileTransferComplete(false);
+      return;
+    }
+
+    // If control reaches here, means handshake was successful
+    selectedPeerDeviceInetAddress = clientAddress;
+    startFileTransfer();
+  }
+
+  private void startFileTransfer() {
+    if (isGroupFormed() && !isFileSender) {
+      ((Callbacks) activity).onFilesForTransferAvailable(filesForTransfer);
+
+      receiverDeviceAsyncTask = new ReceiverDeviceAsyncTask(this);
+      receiverDeviceAsyncTask.execute();
+
+    } else if (isGroupFormed()) { // && isFileSender
```

Comment:
a better expression of this is 
```
if (isGroupFormed){
  if (isFileSender){
  }
  else {
  }
}
```

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/227b5228f8dc09c7ed8351397c3854759516d9a3#diff-7e01f6f425daaca67175b74c839c6af2R338)